poo/abstraction/en/Abstraction.md

```diff
@@ -0,0 +1,131 @@
+
+
+```java
+
+class Human {
+
+    int air;
+    int satiety;
+    float energy;
+
+
+    void breathe() {
+        air += 10;
+    }
+
+    void eat() {
+        satiety += 5;
+    }
+
+    void move() {
+        System.out.println("Un humain souhaite se déplacer !");
+    }
+
+    void consumeEnergy(float value) {
+        this.energie -= value;
+    }
+}
+```
+
+Now, imagine that we have a class `Hiking` that allows us to organise a walk. A thing of style:
+
+```java
+
+class Hiking {
+
+    List<Human> participants;
+
+    public Hiking() { this.participants = new ArrayList<>(); }
+
+    void addParticipant(Human participant) { this.participants.add(participant); }
+
+    void moveAll() {
+        for(Human human : participants) {
+            human.move();
+            human.consumeEnergy(3);
+        }
+    }
+}
+```
+
+So far so good, we have that we want. This code works perfectly, except, in reality, it contains two major problems.
+
+1. **We restraint the hiking of the objects of type Human**, whereas anything can **walk** and **exhausted** could take a hiking. It’s not really the kind to forbid to elephants and to tractors to participate :(
+2. We ask a whole human but us all we need is their legs (hum). This class hiking, however currently access to the human respiratory system (via `breathe()` ) and to his stomach (via `eat()` ) ! **So, we give too much power to this class Hiking, since it is able to do things who shouldn’t be accessible to it**. <br>
+
+So, how to get around this two problems? 
+
+The **interfaces** come to our rescue (interfaces are great). Let's make a `Walking` interface, which allows to describe something that moves and gets tired:
+
+```java
+interface Walking {
+    void move();
+    void consumeEnergy(float value);
+}
+```
+
+So, now, our class Human will be able to **implement** this super interface, and in reality we add `implements Walking` and put a little `@Override` on their methods `move` and `consumeEnergy` and the turn is played, since this is a behaviour that we had already implemented beforehand, just without the interface.
+
+```java
+class Human implements Walking {
+
+    int air;
+    int satiety;
+    float energy;
+
+
+    void breathe() {
+        air += 10;
+    }
+
+    void eat() {
+        satiety += 5;
+    }
+
+    @Override
+    void move() {
+        System.out.println("Un humain souhaite se déplacer !");
+    }
+
+    @Override
+    void consumeEnergy(float value) {
+        this.energy -= value;
+    }
+}
+```
+
+But, what to change? Now, instead of making a human list in our Hiking, we will do a **Walking list** (it’s more politically correct). And, there is almost nothing to change, it’s a good sign:

```

Comment:
```suggestion
But, what has changed? Well, now instead of making a Human list in our Hiking class, we will do a **Walking list** (it’s more politically correct). And, there is almost nothing to change, which is a good sign:

```

Reply:
I think you misunderstood the sentence. Maybe I'm wrong but I can't check for the french version now